### ISSUE TITLE
Feature request: Add setOptions() to web interface..

### ISSUE THREAD

**Issue by rorywalsh** [csound/csound#833]
It would be nice to have to ability to pass command line arguments using setOption with the WASM version of Csound.

**Comment by gogins** [comment]
The WebAssembly interface does not conform as precisely or as broadly as I
would like with the other language APIs, (C, C++, csound.node. Csound for
Android), ctcsound, Java). I am not sure whether working with the existing
code and adding to or modifying it is the best approach, or whether it
would be best to start over and design a WebAssembly interface from scratch
in a new directory, branch, or even repository.

One obvious problem is that the sequence of API calls for the other
language APIs is:

csoundCreate
csoundSetOption
csoundCompile in one form or another
csoundReadScore possibly
csoundStart
csoundPerform

But in WASM it is:

[csoundSetOption not available]
csoundCompile in one form or another
csoundReadScore possibly
csoundStart [does what csoundPerform does on other platfrorms]
[csoundPerform not availale]

The major problem is that csoundStart does one thing on WASM and something
altogether different on all other platforms.

In addition, cross-platform compatibility is broken with respect to the use
of the DOM to identify predefined elements on the Web page that Csound will
use as callbacks or from callbacks.

However, I will look into adding csoundSetOption to the WASM interface.

Regards,
Mike





-----------------------------------------------------
Michael Gogins
Irreducible Productions
http://michaelgogins.tumblr.com
Michael dot Gogins at gmail dot com

**Comment by vlazzarini** [comment]
done. There is a CsoundObj.setOption(string) now in the interface.

**Comment by rorywalsh** [comment]
Thanks Victor. Mike, I agree it's not ideal, but it presents the path of
least resistance for developers, which as you know is often paramount in
open source development. I'm very happy and thankful to have this, even if
it's a little different to the other interfaces. Thanks again to all of you
for you work on it.

**Comment by gogins** [comment]
Victor, before it is too late, in the WASM API can we rename csound.start()
to csound.perform() and put in a minimal or stub csound.start() to be
consistent with the other APIs? This of course would apply to examples as
well.

I think it would also be a good idea to protect against contexts that lack
the expected callback functions that the module hooks into, either with
exception handling, or by installing callbacks in the document if they
don't already exist.

Again, the goal is for a minimal subset of the Csound API to be named the
same, and behave the same, across all platforms.

Best,
Mike


-----------------------------------------------------
Michael Gogins
Irreducible Productions
http://michaelgogins.tumblr.com
Michael dot Gogins at gmail dot com

**Comment by vlazzarini** [comment]
I'd like to ask steven @kunstmusik, since the design is originally his. I am happy either way.

**Comment by hlolli** [comment]
@vlazzarini found a typeo
https://github.com/csound/csound/blob/develop/emscripten/build.sh#L21
'CsoundObj_setOption'
should be
'_CsoundObj_setOption'

Otherwise this function doesn't appear in libcsound.js

**Comment by hlolli** [comment]
Also looks like `_CsoundObj_compileOrc` inheritely calls csoundStart because when I call it the second time I get, could be a different reason tough.
```
Csound is already started, call csoundReset()
before starting again 
255
```

**Comment by vlazzarini** [comment]
Are you calling _CsoundObj_compileOrc() directly? In any case you should always call
the methods  from the CsoundObj class instead
of the wrapped C functions. 
It might not solve your particular issue, but it will save problems later.

**Comment by hlolli** [comment]
Yup, that was the thing, I didn't realize that the play in the examples is a wrapper of .CompileOrc and .CsoundStart. But I'm running on nodejs on a different programming language, so I had to brake this a part into smaller units, crwap didn't run well with node and I need to connect to different audio engine. So I rewrote the start function with the node wrappers
```Clojure
(defn wasm-start [csound-object csound-instance]
  (.setFlagsFromString (js/require "v8") "--no-use_strict") ;; To be able to load web-audio-api
  (let [ksmps (._CsoundObj_getKsmps csound-object csound-instance)
        input-count (._CsoundObj_getInputChannelCount csound-object csound-instance)
        output-count (._CsoundObj_getOutputChannelCount csound-object csound-instance)
        audio-context-constructor (.-AudioContext (js/require "web-audio-api"))
        audio-context (new audio-context-constructor)
        audio-process-node (.createScriptProcessor
                            audio-context
                            1024 input-count output-count)
        buffer-size (.-bufferSize audio-process-node)
        output-pointer (._CsoundObj_getOutputBuffer csound-object csound-instance)
        csound-output-buffer (new js/Float32Array (.-buffer (.-HEAP8 csound-object))
                                  output-pointer (* ksmps output-count))
        ;; TODO add microphone input buffer
        zerodbfs (._CsoundObj_getZerodBFS csound-object csound-instance)
        process-buffers (fn [e sample-count src-offset dst-offset]
                          (for [i (range output-count)]
                            (for [j (range sample-count)]
                              (-> (.-inputBuffer e )
                                  (.getChannelData i)
                                  (aset (+ j src-offset)
                                        (-> csound-output-buffer
                                            (aget (+ i (* output-count
                                                          (+ j src-offset))))))
                                  (/ zerodbfs)))))
        offset (volatile! ksmps)
        Speaker (js/require "speaker")]
    (set! (.-outStream audio-context)
          (new Speaker #js {:channels (.-numberOfChannels
                                       (.-format audio-context))
                            :bitDepth (.-bitDepth
                                       (.-format audio-context))
                            :sampleRate (.-sampleRate audio-context)}))
    (.connect audio-process-node (.-destination audio-context))
    (set! (.-onaudioprocess audio-process-node)
          (fn [e]
            (let [offs (:val offset)
                  sample-count (- ksmps offs)
                  index (if (< 0 sample-count)
                          (do (process-buffers e sample-count offs 0) sample-count)
                          0)]
              (loop [index index
                     sample-count sample-count]
                (if-not (< index buffer-size)
                  (vreset! offset sample-count)
                  (let [res (._CsoundObj_performKsmps csound-object csound-instance)
                        sample-count (Math/min ksmps (- buffer-size index))]
                    (if (not= 0 res)
                      (do (.disconnect audio-process-node)
                          (set! (.-onaudioprocess audio-process-node) nil))
                      (do (when (js/isNaN (aget csound-output-buffer 0))
                            (.error js/console (str "NaN! outputPointer = " output-pointer)))
                          (process-buffers e sample-count 0 index)
                          (recur (+ index sample-count)
                                 sample-count)))))))))
    nil))
```
And with this interface
```Clojure
(deftype CsoundWASM [csound-object csound-instance]
  CsoundInterface
  (compile-orc [this orc]
    (._CsoundObj_evaluateCode
     csound-object csound-instance orc))
  (input-message [this input-message]
    (._CsoundObj_readScore csound-object csound-instance input-message))
  (reset [this] (._CsoundObj_reset csound-object csound-instance))
  (play [this] (._CsoundObj_play csound-object csound-instance))
  (set-option [this option] (._CsoundObj_setOption csound-object csound-instance option))
  (start [this]
    (._CsoundObj_compileOrc csound-object csound-instance "nchnls=2\n 0dbfs=1\n")
    (wasm-start csound-object csound-instance)))
```
Calling it in this order
```Clojure
(create)

0dBFS level = 32768.0
--Csound version 6.09.1 (double samples) Aug  3 2017 
[commit: b67f555e5bc84c9264b784d47121c1fa8a8978e9]
libsndfile-1.0.25

(start csound)
 
Failed to open input file (null)

graphics suppressed, ascii substituted
0dBFS level = 32768.0
orch now loaded
audio buffered in 256 sample-frame blocks
writing 512-byte blks of shorts to test.wav (WAV)
SECTION 1:
nil

(compile-orc csound
             (str "instr 1\n"
                  "a1 poscil 0.9, 200\n"
                  "out a1\n"
                  "endin"))
 
NaN

(input-message csound "e 0 3600") 

0

(input-message csound "i 0 10")
 
  rtevent:	   T  3.701 TT  3.701 M:      0.0
Score finished in csoundPerformKsmps() with 2.
0
```

Just a braindump that I don't recommend reading too thoroughly.

**Comment by vlazzarini** [comment]
Not sure what you are doing. csoundOption() is a method of the JS class CsoundObj.

https://github.com/csound/csound/blob/develop/emscripten/src/CsoundObj.js#L162

_CsoundObj_setOption() is the C function exported to JS. You should not call that directly.

Also _CsoundObj_compileOrc()  only calls csoundStart() if the internal status of the CsoundObj is reset.

**Comment by vlazzarini** [comment]
the best thing is probably to provide a lower-level interface for these types of things in CsoundObj.

**Comment by hlolli** [comment]
Yes, I am bypassing CsoundObj entirely and makeing my own with protocols and types in Clojurescript. So there's already good as complete interface, how they are combined makes CsoundObj.js and csound.js a good documentaion source.

About setOption,  csoundOption()  inside CsoundObj.js calls '_CsoundObj_setOption' but wont succeed due to the typo in the bash script I pointed out, otherwise I see that I have no need for the option to solve the problems Im faceing, which I think is mostly due to the ordering of startup event. I think with one more coffee or sleep I will get this running.

**Comment by kunstmusik** [comment]
@hlolli In case it's of any use, I have Clojurescript wrapper to CsoundObj that optionally loads asm.js or wasm:

https://github.com/kunstmusik/csound-notebook/blob/clj/src/cljs/csound_notebook/csound.cljs

**Comment by kunstmusik** [comment]
@gogins @vlazzarini I'm happy to go with breaking API changes as long as it's documented in the ChangeLog and put into the release announcement. If we're going to do an API change for consistency though, we should look at the Android and iOS CsoundObj interfaces too.  I'd personally rather we do all of this in Csound 7 though.

**Comment by vlazzarini** [comment]
fixed the typo. I was looking at the wrong script (build.sh).

**Comment by hlolli** [comment]
@kunstmusik very nice, but the difference is that I'm compiling to node.js rather than the browser. But that adventurous quest is getting tiresome, for very event I send I always end up getting 
```
Score finished in csoundPerformKsmps() with 2.
```
I can run performKsmps many times, but as soon as event hits that loop, even if it's an score event of "e 0 3600", it just gives me Score finished. What can be the cause of this, the api or could reading the audio buffers be in any way connected to it? (it happens with or without reading from the output buffers).

**Comment by hlolli** [comment]
This sample is also broken?
https://csound.github.io/wasm/minimal.html

WARNING: [mCsound not ready for performance: csoundStart() has not been called 
[m

**Comment by vlazzarini** [comment]
maybe the changes I made yesterday did this. I will need to review all of these.

**Comment by hlolli** [comment]
ok nice, no stress.

**Comment by vlazzarini** [comment]
that's fixed now. Some commits had broken csound.js and so I had to spend time finding what had been changed and fixing it.

**Comment by hlolli** [comment]
https://csound.github.io/wasm/minimal.html

Is it me or do you also get "Aw, Snap!" screen on Chrome?

**Comment by vlazzarini** [comment]
works here.
========================
Prof. Victor Lazzarini
Dean of Arts, Celtic Studies, and Philosophy,
Maynooth University,
Maynooth, Co Kildare, Ireland
Tel: 00 353 7086936
Fax: 00 353 1 7086952

**Comment by hlolli** [comment]
Ok updateing from Chrome58 to Chrome60 fixes it (but the example is still the same as last time).

**Comment by vlazzarini** [comment]
the example is fixed. You need to make sure the latest csound.js is being used (from emscripten/src/csound.js)
https://github.com/csound/csound/blob/develop/emscripten/src/csound.js

========================
Prof. Victor Lazzarini
Dean of Arts, Celtic Studies, and Philosophy,
Maynooth University,
Maynooth, Co Kildare, Ireland
Tel: 00 353 7086936
Fax: 00 353 1 7086952

**Comment by hlolli** [comment]
Yup it works if I run it locally, just need to add /js/ prefix to the `var wasmBinaryFile = Module['wasmBinaryFile'] || './js/libcsound.wasm';`
The speaker gets a bit juiced up with this console.error
`NaN! outputPointer = 6633640`

**Comment by hlolli** [comment]
Finally! I'm able to run wasm csound in node successfully, mistakenly took cwrap to be a library and by using that everything works.

I'm getting sound but I'm hitting the exact same NaN! bug as with the minimal example
```
panaeolus.engine=>   rtevent:	   T 13.770 TT 13.770 M:  0.89999  0.89999
NaN! outputPointer = 6633640
NaN! outputPointer = 6633640
NaN! outputPointer = 6633640
```

**Comment by vlazzarini** [comment]
Since this unrelated to the original bug, and in the interest of the community, can we move this discussion to the developers' list?

**Comment by hlolli** [comment]
Yes
